// transition-group/dist/TransitionGroup.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactAddonsTransitionGroup = require('react-addons-transition-group');

var _reactAddonsTransitionGroup2 = _interopRequireDefault(_reactAddonsTransitionGroup);

var _domUtils = require('./dom-utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var lastUpdate = new Date();

var TransitionGroup = function (_React$Component) {
  _inherits(TransitionGroup, _React$Component);

  function TransitionGroup() {
    _classCallCheck(this, TransitionGroup);

    return _possibleConstructorReturn(this, (TransitionGroup.__proto__ || Object.getPrototypeOf(TransitionGroup)).apply(this, arguments));
  }

  _createClass(TransitionGroup, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (this.props.onEmpty && _react2.default.Children.count(nextProps.children) === this.props.zeroElements && _react2.default.Children.count(this.props.children) > this.props.zeroElements) {
        var _duration = (this.props.leaveDuration || this.props.duration) + (this.props.leaveDelay || this.props.delay);
        (0, _domUtils.setTimeoutAnimationFrame)(this.props.onEmpty, _duration);
      }

      if (this.props.onFull && _react2.default.Children.count(nextProps.children) > this.props.zeroElements && _react2.default.Children.count(this.props.children) === this.props.zeroElements) {
        (0, _domUtils.setTimeoutAnimationFrame)(this.props.onFull, 0);
      }
    }

    // simple debounce mechanism that works

  }, {
    key: 'shouldComponentUpdate',
    value: function shouldComponentUpdate(nextProps) {
      var _this2 = this;

      var shouldDebounce = _react2.default.Children.count(nextProps.children) === 1 && _react2.default.Children.count(this.props.children) === 1;

      if (!shouldDebounce) return true;

      var latestChange = new Date();
      var delta = latestChange - lastUpdate;

      if (delta < (nextProps.duration || 0)) {
        setTimeout(function () {
          return _this2.setState({});
        }, delta + 34); // 34 is 2 animation frames just like the animations do
        return false;
      }

      lastUpdate = new Date();
      return true;
    }
  }, {
    key: 'render',
    value: function render() {
      var _props = this.props,
          children = _props.children,
          prefix = _props.prefix,
          duration = _props.duration,
          delay = _props.delay,
          appear = _props.appear,
          enter = _props.enter,
          leave = _props.leave,
          enterDuration = _props.enterDuration,
          leaveDuration = _props.leaveDuration,
          appearDuration = _props.appearDuration,
          enterDelay = _props.enterDelay,
          leaveDelay = _props.leaveDelay,
          appearDelay = _props.appearDelay,
          onAppear = _props.onAppear,
          onEnter = _props.onEnter,
          onLeave = _props.onLeave,
          onBeforeAppear = _props.onBeforeAppear,
          onBeforeEnter = _props.onBeforeEnter,
          onBeforeLeave = _props.onBeforeLeave,
          onEmpty = _props.onEmpty,
          onFull = _props.onFull,
          zeroElements = _props.zeroElements,
          props = _objectWithoutProperties(_props, ['children', 'prefix', 'duration', 'delay', 'appear', 'enter', 'leave', 'enterDuration', 'leaveDuration', 'appearDuration', 'enterDelay', 'leaveDelay', 'appearDelay', 'onAppear', 'onEnter', 'onLeave', 'onBeforeAppear', 'onBeforeEnter', 'onBeforeLeave', 'onEmpty', 'onFull', 'zeroElements']);

      return _react2.default.createElement(
        _reactAddonsTransitionGroup2.default,
        props,
        _react2.default.Children.map(children || [], function (child) {
          if (!child) return null; // cloneElement won't work on a non-existent child (null will filter it out)

          return _react2.default.cloneElement(child, _extends({
            prefix: prefix,
            duration: duration,
            delay: delay,

            appear: appear,
            enter: enter,
            leave: leave,

            enterDuration: enterDuration,
            leaveDuration: leaveDuration,
            appearDuration: appearDuration,

            enterDelay: enterDelay,
            leaveDelay: leaveDelay,
            appearDelay: appearDelay,

            onAppear: onAppear,
            onEnter: onEnter,
            onLeave: onLeave,

            onBeforeAppear: onBeforeAppear,
            onBeforeEnter: onBeforeEnter,
            onBeforeLeave: onBeforeLeave
          }, child.props));
        }).filter(function (child) {
          return child;
        })
      );
    }
  }]);

  return TransitionGroup;
}(_react2.default.Component);

TransitionGroup.defaultProps = {
  prefix: '',
  duration: 0,
  delay: 0,

  appear: 'appear',
  enter: 'enter',
  leave: 'leave',

  appearDuration: 0,
  enterDuration: 0,
  leaveDuration: 0,

  appearDelay: 0,
  enterDelay: 0,
  leaveDelay: 0,

  zeroElements: 0
};
TransitionGroup.propTypes = {
  children: require('prop-types').any,
  prefix: require('prop-types').string,
  duration: require('prop-types').number,
  delay: require('prop-types').number,
  appear: require('prop-types').string,
  enter: require('prop-types').string,
  leave: require('prop-types').string,
  appearDuration: require('prop-types').number,
  enterDuration: require('prop-types').number,
  leaveDuration: require('prop-types').number,
  appearDelay: require('prop-types').number,
  enterDelay: require('prop-types').number,
  leaveDelay: require('prop-types').number,
  onBeforeAppear: require('prop-types').func,
  onBeforeEnter: require('prop-types').func,
  onBeforeLeave: require('prop-types').func,
  onAppear: require('prop-types').func,
  onEnter: require('prop-types').func,
  onLeave: require('prop-types').func,
  onEmpty: require('prop-types').func,
  onFull: require('prop-types').func,
  zeroElements: require('prop-types').number
};
exports.default = TransitionGroup;
